### PR TITLE
Liquids system

### PR DESCRIPTION
**Pull request by KayeArray**
:cl: banthebantz
add: Chempiles have been replaced with liquids, they flow, have depth, have elevation simulation etc
add: Drains have been added to the map, these will drain liquids at a rate based on their viscosity
/:cl:

[why]: Goon gave me the idea

Yeah so basically the map merge doesn't work so enjoy the diffs

https://www.youtube.com/watch?v=wlzFQBvdHH4&feature=youtu.be

### CODE REVIEW

**Comment by ghost** [comment]
hippie is living in the 4017ies now

File: hippiestation/code/game/objects/liquid.dm [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+			total_activity = 0
+			counter = 0
+		if(average_viscosity && LAZYLEN(liquids))//fucking division by zero shit
+			average_viscosity = max(average_viscosity / LAZYLEN(liquids), 0.1)
+		spread_time = world.time + throttle + average_viscosity
+
+
+/obj/effect/liquid
+	name = "Liquid"
+	desc = "Looks wet"
```
Comment (praisenarsie) [inline hippiestation/code/game/objects/liquid.dm]: decapitalize liquid and add a period at the end of the desc.

File: hippiestation/code/modules/reagents/chemistry/reagents/toxin_reagents.dm [hunk 2]
```diff
@@ -436,4 +436,19 @@
 			var/mob/living/carbon/human/H = M
 			H.disabilities |= CLUMSY
 
+	..()
```
Comment (praisenarsie) [inline hippiestation/code/modules/reagents/chemistry/reagents/toxin_reagents.dm:439]: is this supposed to be here or am i missing something?

File: hippiestation/code/modules/reagents/chemistry/reagents/toxin_reagents.dm [hunk 3]
```diff
@@ -436,4 +436,19 @@
...
+
+/datum/reagent/toxin/vomit
+	name = "Vomit"
+	id =  "vomit"
+	description = "Chunky"
```
Comment (praisenarsie) [inline hippiestation/code/modules/reagents/chemistry/reagents/toxin_reagents.dm]: . period and there's a double space on the ``=  "vomit"``

File: hippiestation/code/game/objects/liquid.dm [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+	alpha = 100
+	mouse_opacity = MOUSE_OPACITY_TRANSPARENT
+	layer = LOW_OBJ_LAYER
+	resistance_flags = LAVA_PROOF | FIRE_PROOF | UNACIDABLE | ACID_PROOF
+	var/viscosity = 0//affects the spread and general properties of the liquid
+	var/depth = 0//how much liquid is on this tile
+	var/volatile = FALSE//does it evaporate on its own?
+	var/spread_rate = 1//self explanatory
+	var/is_static = FALSE//a static liquid will never lose volume and will only add to other liquids, good for permanent liquid sources
+	var/datum/liquid_pool/pool///a pool is a group of interconnected liquid tiles that process together, this is used for organisation and optimisation
```
Comment (praisenarsie) [inline hippiestation/code/game/objects/liquid.dm]: missing space for the comment

File: hippiestation/code/game/objects/liquid.dm [hunk 5]
```diff
@@ -0,0 +1,276 @@
+#define LIQUID_TICKS_UNTIL_THROTTLE 50
+#define REAGENT_TO_DEPTH 2//one 'depth' per 2u
+#define MAX_INITIAL_DEPTH 25
+#define LERP(a, b, amount) (amount ? (a + (b - a) * amount) : (a + (b - a) * 0.5))
+
+/datum/liquid_pool//abstract shit to manage pools of liquid
+	var/total_activity//cached activity for last 50 runs
```
Comment (praisenarsie) [inline hippiestation/code/game/objects/liquid.dm]: missing space for the comment

File: hippiestation/code/game/objects/effects/decals/cleanable/misc.dm [hunk 6]
```diff
@@ -0,0 +1,6 @@
+/obj/effect/decal/cleanable/vomit/Initialize()
+	. =..()
```
Comment (praisenarsie) [inline hippiestation/code/game/objects/effects/decals/cleanable/misc.dm]: ``. = ..()``

File: hippiestation/code/game/objects/structures/drain.dm [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	density = FALSE
+	anchored = TRUE
+	layer = GAS_SCRUBBER_LAYER
+	max_integrity = 400//it's small and made of thick strong metal
+	var/can_weld_shut = TRUE
+	var/datum/looping_sound/drain/soundloop
+	var/counter = 0
+
+/obj/structure/drain/Initialize()
+	. =..()
```
Comment (praisenarsie) [inline hippiestation/code/game/objects/structures/drain.dm:15]: ``. = ..()``

**Comment by GuyonBroadway** [comment]
FIGGS DE CONFLIGDS

**Comment by ghost** [comment]
replace the pool with liquid water

**Comment by GuyonBroadway** [comment]
Poolcode is shitcode and needs a massive refactor. I'd say its beyond the scope of this PR but I'll see what I can do about it later.

**Comment by ghost** [comment]
also, @banthebantz, idea: Keep chempiles but make them only happen if you splash a low amount of a chem

File: hippiestation/code/game/objects/liquid.dm [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+	var/cached_activity = 0//this is used to judge the activity of a pool, if it is 0 or close to 0 the processing for a pool will be throttled or stopped to save performance
+
+
+/obj/effect/liquid/Initialize()
+	. = ..()
+	create_reagents(1000)
+	addtimer(CALLBACK(src, .proc/get_pool), 0)
+
+
+/obj/effect/liquid/proc/get_pool()
```
Comment (ghost) [inline hippiestation/code/game/objects/liquid.dm]: I have a feeling this is going to create millions of datums

**Comment by ghost** [comment]
Fix the DMM so this will compile

**Comment by ghost** [comment]
you are an amazing man were i gay id hug you

**Comment by ghost** [comment]
make it so when water moves onto a tile, it can carry TINY or SMALL items with it
my immersions

**Comment by KayeArray** [comment]
water does push stuff when it moves

File: hippiestation/code/game/objects/liquid.dm (hunk 8 above)
Comment (KayeArray) [reply hippiestation/code/game/objects/liquid.dm]: not really

**Comment by dreamcrab** [comment]
this is some cool og kush india weed shit 

+840

**Comment by ghost** [comment]
if is cool +1

File: hippiestation/code/modules/reagents/chemistry/reagents/toxin_reagents.dm (hunk 2 above)
Comment (KayeArray) [reply hippiestation/code/modules/reagents/chemistry/reagents/toxin_reagents.dm:439]: yes

**Comment by ghost** [comment]
amazing

**Comment by ghost** [comment]
The future is now +1

**Review by GuyonBroadway — APPROVED** [review]
this looks ok to me, shame about the map crap but it happens.

**Comment by GuyonBroadway** [comment]
Headmin Approved, Travis approved coder approved. 

OPEN THE FLOODGATES!!

**Comment by Manly-Dwarf** [comment]
Holy fuck did anyone think this through?

**Comment by ghost** [comment]
probably not

**Comment by msgerbs** [comment]
>not testmerging something this big

Why tho?

**Pull request merged**